packages/multichain/src/scope/transform.ts

```diff
@@ -0,0 +1,145 @@
+import type { CaipReference } from '@metamask/utils';
+import { cloneDeep } from 'lodash';
+
+import type {
+  ExternalScopeObject,
+  ExternalScopesObject,
+  InternalScopeString,
+  InternalScopeObject,
+  InternalScopesObject,
+} from './types';
+import { parseScopeString } from './types';
+
+/**
+ * Returns a list of unique items
+ *
+ * @param list - The list of items to filter
+ * @returns A list of unique items
+ */
+export const getUniqueArrayItems = <Value>(list: Value[]): Value[] => {
+  return Array.from(new Set(list));
+};
+
+/**
+ * Normalizes a ScopeString and ExternalScopeObject into a separate
+ * InternalScopeString and InternalScopeObject for each reference in the `references`
+ * value if defined and adds an empty `accounts` array if not defined.
+ *
+ * @param scopeString - The string representing the scope
+ * @param externalScopeObject - The object that defines the scope
+ * @returns a map of caipChainId to ScopeObjects
+ */
+export const normalizeScope = (
+  scopeString: string,
+  externalScopeObject: ExternalScopeObject,
+): InternalScopesObject => {
+  const { references, ...scopeObject } = externalScopeObject;
+  const { namespace, reference } = parseScopeString(scopeString);
+
+  const normalizedScopeObject: InternalScopeObject = {
+    accounts: [],
+    ...scopeObject,
+  };
+
+  const shouldFlatten =
+    namespace &&
+    !reference &&
+    references !== undefined &&
+    references.length > 0;
+
+  if (shouldFlatten) {
+    return Object.fromEntries(
+      references.map((ref: CaipReference) => [
+        `${namespace}:${ref}`,
+        cloneDeep(normalizedScopeObject),
+      ]),
+    );
+  }
+  return { [scopeString]: normalizedScopeObject };
+};
+
+export const mergeScopeObject = (
```

Comment:
Nit: A TSDoc entry would be helpful here, especially given the very-similar method `mergeScopes` that could be confused with this

Reply:
done here: https://github.com/MetaMask/core/pull/4784/commits/5633bddd0a517ab5549d144281a0e17d18ab4f3f